### DISCUSSION TITLE
CoolProp dynamic viscosity and speed of sound models for hydrogen

### DISCUSSION THREAD

**Discussion by Nicolas-Capmany** [CoolProp/CoolProp#2370] [Q&A]
Hi Dr Bell, 

I have been recently looking into using CoolProp for modelling the different nuclear isomers of hydrogen: 

- Parahydrogen
- Orthohydrogen
- Normal hydrogen (3:1 ortho-para composition)

However, I have come across the following details concerning **dynamic viscosity models**:

| Parahydrogen | Orthohydrogen | Normal hydrogen |
|----------------|-------------------|---------------------|
| Model unavailable| Model unavailable | Model available |

and with respect to the **speed of sound models**:

| Parahydrogen | Orthohydrogen | Normal hydrogen |
|----------------|-------------------|---------------------|
| Model available| Model available | Model unavailable |

I was wondering if the speed of sound models and dynamic models missing were deliberately omitted or not, and for those that 
are present (e.g., dynamic viscosity for normal hydrogen and speed of sound models for para- and ortho- hydrogen), would it be possible to know which models where implemented?

Thank you very much for your time.

**Comment by ibell** [comment]
To start, recommend to check out the references for the implemented models: http://www.coolprop.org/fluid_properties/PurePseudoPure.html#list-of-fluids

I have to say I am surprised that the speed of sound does not work for normal H2. What error do you get? All the parts are implemented to allow for those calculations.

**Comment by Nicolas-Capmany** [comment]
Thank you very much for your quick response Dr Bell, 

Concerning the speed of sound for normal Hydrogen, I tried the following code to check if I could get it, but I bumped into problems:

```python
# import libraries
from CoolProp.CoolProp import PropsSI
import numpy as np

# initialise density and temperature arrays
rho     = np.full(5, 69.175)
T       = np.full(5, 21.774)
P       = PropsSI("P", "D", rho, "T", T, "Hydrogen") # this yields a pressure ~ 150000 Pa

# Calculate speed of sound for normal hydrogen
a_d = PropsSI("A", "D", rho, "P", P, "Hydrogen")
```

However, if I write another code in which I manually prescribe a pressure field of ~`245000 Pa` CoolProp is actually able to yield a speed of sound value:

```python
# import libraries
from CoolProp.CoolProp import PropsSI
import numpy as np

# initialise density and temperature arrays
rho     = np.full(5, 69.175)
P        = np.full(5, 245000) # Manually imposing a pressure value of 245000 Pa

# Calculate the speed of sound for normal hydrogen
a_d = PropsSI("A", "D", rho, "P", P, "Hydrogen")
```

In this case, this pressure field can yield a value for the speed of sound. I was wondering, am I missing something from the thermodynamics side? Is it possible there might be an error with the pressure field of normal hydrogen when working in the lower end of the temperature spectrum?

Thank you very much for your time.

**Comment by ibell** [comment]
You should provide single values to PropsSI when debugging as that will give you a stack trace on the error. In this case, an informative error message:
```python
# import libraries
from CoolProp.CoolProp import PropsSI
import numpy as np

# initialise density and temperature arrays
P = PropsSI("P", "D", 69.175, "T", 21.774, "Hydrogen") # this yields a pressure ~ 150000 Pa

# Calculate speed of sound for normal hydrogen
a_d = PropsSI("A", "D", 69.175, "P", P, "Hydrogen")
```
yielding
```
Exception has occurred: ValueError
Speed of sound is not defined for two-phase states because it depends on the distribution of phases. : PropsSI("A","D",69.175,"P",149982.5145,"Hydrogen")
  File "C:\Users\ian\Desktop\script.py", line 9, in <module>
    a_d = PropsSI("A", "D", 69.175, "P", P, "Hydrogen")
ValueError: Speed of sound is not defined for two-phase states because it depends on the distribution of phases. : PropsSI("A","D",69.175,"P",149982.5145,"Hydrogen")
```
